File: workers/javascript/packages/server/src/store/redis.ts
import { RedisClientType } from '@redis/client';
import { RedisCommandRawReply } from '@redis/client/dist/lib/commands';
import { MaybeError } from '@superblocks/shared';
import { Logger } from 'pino';
import { QuotaError } from '../errors';
import { IO, KV, KVOps, KVStore, KVStoreTx, RedisOptions, Wrapped, WriteOps } from '../types';

export class RedisTx implements KVStoreTx {
  /**
   * DEFAULTS
   */
  private readonly MAX_BYTES = 500000000;
  private readonly SLOW_THRESHOLD = 100000000;

  private _ops: { op: 'read' | 'write'; key: string; value?: string; expiration?: number }[];
  private _fastClient: RedisClientType;
  private _slowClient: RedisClientType;
  private _ttl?: number;
  private _logger?: Logger;
  private _maxBytes: number;
  private _slowThreshold: number;
  private _bytes: number;
  private _committed: boolean;

  constructor(options: RedisOptions) {
    this._ops = [];
    this._fastClient = options.client;
    this._slowClient = options.slowClient || this._fastClient;
    this._logger = options.logger?.child({ who: 'kvstore' });
    this._maxBytes = options.maxBytes || this.MAX_BYTES;
    this._slowThreshold = options.slowThreshold || this.SLOW_THRESHOLD;
    this._committed = false;
    this._bytes = 0;

    if (options.seconds > 0) {
      this._ttl = options.seconds;
    }
  }

  public read(key: string): void {
    if (key === '') {
      throw Error('key must be defined');
    }

    this._ops.push({ op: 'read', key });
  }

  public write(key: string, value: unknown, ops?: WriteOps): void {
    if (key === '') {
      throw Error('key must be defined');
    }

    // NOTE(frank): This scares me. I'm not sure what the side effects are of this.
    // NOTE(dlamotte): previous check was !value, which when dropped led to the following error:
    //   The "string" argument must be of type string or an instance of Buffer or ArrayBuffer. Received undefined
    if (value === undefined) {
      return;
    }

    const encoded: string = JSON.stringify(value);
    const bytes: number = Buffer.byteLength(encoded);

    /**
     * QUOTA ENFORCEMENT
     */
    if (ops?.maxSize && bytes > ops.maxSize) {
      this._logger?.warn({ ...ops?.baggage, ...{ size: bytes, limit: ops.maxSize } }, "This value's size has exceeded the limit.");
      throw new QuotaError(`This value's size (${bytes}) has exceed the limit (${ops.maxSize}).`);
    }

    if (bytes > this._maxBytes) {
      // NOTE(frank): Should we clear out the ops? Should we add an abort() method an call it to cleanup?
      throw new Error(
        `The serialized size of the output (${Math.ceil(bytes / 1000000)}MB) exceeds to maxixum (${Math.floor(
          this._maxBytes / 1000000
        )}MB).`
      );
    }

    this._bytes += bytes;
    this._ops.push({ op: 'write', key, value: encoded, expiration: ops?.expiration || this._ttl });
  }

  public async commit(ops?: KVOps): Promise<Wrapped<IO, Array<unknown>>> {
    if (this._committed) {
      throw Error('this transaction has already been committed');
    }

    const io: IO = {};

    if (this._ops.length < 1) {
      return { pinned: io, data: [] };
    }

    io.write = this._bytes;

    let tx = (this._bytes > this._slowThreshold ? this._slowClient : this._fastClient).multi();
    const reads: string[] = [];
    const writes: string[] = [];

    this._ops.forEach(({ op, key, value, expiration }): void => {
      switch (op) {
        case 'read':
          reads.push(key);
          tx = tx.get(key);
          break;
        case 'write':
          writes.push(key);
          tx = tx.set(key, value, expiration ? { EX: expiration } : {});
          break;
        default:
          return;
      }
    });

    this._logger?.debug({ ...ops?.baggage, ...{ io, reads: reads.length, writes: writes.length } }, 'before committing transaction');

    const data: Array<unknown> = [];

    let results: RedisCommandRawReply;
    {
      try {
        results = await tx.exec();
      } catch (err) {
        this._logger?.error(
          {
            ...ops?.baggage,
            ...{ io, reads: reads.length, writes: writes.length, err }
          },
          'error committing transaction'
        );
        throw Error('could not commit transaction');
      } finally {
        this._committed = true;
      }
    }

    if (results.length != this._ops.length) {
      throw Error(`Executed ${this._ops.length} ops but only received ${results.length} results.`);
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    results.forEach((result: any, idx: number): void => {
      switch (this._ops[idx].op) {
        case 'read':
          if (!result) {
            // This is the 'null' or 'undefined' case.
            io.read = (io.read || 0) + 1;
          } else {
            try {
              io.read = (io.read || 0) + Buffer.byteLength(result);
            } catch (err) {
              this._logger?.error(`could not determine size of response due to a type issue (${typeof result})`);
            }
          }
          data.push(JSON.parse(result));
          break;
        case 'write':
          if (result != 'OK') {
            // NOTE(frank): Perhaps we don't want to throw as we're swalling the successful read ops.
            throw Error(`There was an issue writing ${this._ops[idx].key}=${this._ops[idx].value}`);
          }
          break;
        default:
          return;
      }
    });

    this._logger?.debug({ ...ops?.baggage, ...{ io, reads: reads.length, writes: writes.length } }, 'committed transaction');

    return { pinned: io, data };
  }
}

export class Redis implements KVStore {
  private _options: RedisOptions;
  private _client: RedisClientType;

  constructor(options: RedisOptions) {
    this._options = options;
    this._client = options.client;
  }

  public async decr(key: string): Promise<number> {
    return await this._client.decr(key);
  }

  public tx(): KVStoreTx {
    return new RedisTx(this._options);
  }

  public async read(keys: string[]): Promise<Wrapped<IO, Array<unknown>>> {
    const tx: KVStoreTx = new RedisTx(this._options);

    keys.forEach((key: string): void => {
      tx.read(key);
    });

    return await tx.commit();
  }

  public async write(key: string, value: unknown, ops: WriteOps = {}): Promise<Wrapped<IO, void>> {
    return this.writeMany([{ key, value, expiration: ops?.expiration }]);
  }

  public async writeMany(payload: KV[], ops: WriteOps = {}): Promise<Wrapped<IO, void>> {
    const tx: KVStoreTx = new RedisTx(this._options);

    payload.forEach(({ key, value, expiration }: KV): void => {
      tx.write(key, value, { ...ops, ...{ expiration } });
    });

    return { pinned: (await tx.commit(ops)).pinned, data: undefined };
  }

  public async delete(keys: string[]): Promise<void> {
    if (keys.length === 0) {
      return;
    }

    try {
      const resp = await this._client.del(keys);
      if (typeof resp !== 'number') {
        throw new Error(`Unexpected response while deleting redis keys: ${resp}.`);
      }
    } catch (err) {
      this._options.logger?.error({ err, keys }, 'Failed to write to redis.');
      throw err;
    }
  }

  public async close(reason?: string): Promise<MaybeError> {
    try {
      await this._client.disconnect();
    } catch (err) {
      this._options.logger?.error({ err, reason }, 'Failed to close redis connection.');
      return err;
    }
  }
}

File: workers/python/src/kvstore/redis.py
from typing import Any, Dict, List, Optional, Tuple

import redis
import ujson
from redis.asyncio import Redis

import superblocks_json
from exceptions import QuotaError
from kvstore.kvstore import KV, KVStore, WriteOps
from log import error, info, warn


class RedisKVStore(KVStore):
    def __init__(self, url="", key_retention_seconds=3600, redis_options: dict = {}):
        if url:
            self._client: Redis = Redis.from_url(url, **redis_options)
        else:
            self._client = Redis(**redis_options)  # type: ignore
        self._key_retention_seconds = key_retention_seconds

    async def read(self, keys: List[str]) -> Tuple[List[Dict], int]:
        if not keys:
            return [], 0

        try:
            values = await self._client.mget(keys)
            result = []
            size_in_bytes = 0
            for idx, value in enumerate(values):
                if value:
                    # Bytstrings are not properly included in the typing lib...
                    result.append(superblocks_json.loads(value.decode()))  # type: ignore
                    size_in_bytes += len(value)
                else:
                    result.append({})
                    try:
                        empty_key = keys[idx]
                    except IndexError:
                        empty_key = "Unknown"
                        error("redis returned an unexpected response when reading keys")

                    warn(f"attempted to read empty key from redis: {empty_key}")
        except Exception as e:
            error(f"Failed to get keys from redis: {e}")
            raise e
        return result, size_in_bytes

    async def write(
        self, key: str, value: Any, expiration_seconds: Optional[int] = None
    ) -> int:
        ttl = expiration_seconds if expiration_seconds else self._key_retention_seconds
        try:
            string_val = ujson.dumps(value)
        except Exception as e:
            error(f"Failed to parse value: {str(e)}")
            raise e

        try:
            if self._key_retention_seconds != -1:
                is_success: Optional[bool] = await self._client.setex(
                    key, ttl, string_val
                )
            else:
                is_success = await self._client.set(key, string_val)

            if not is_success:
                raise Exception(
                    "Attempt to write to redis resulted in a NON-OK response."
                )
        except Exception as e:
            error("failed to write to redis", err=str(e))
            raise e

        return len(string_val.encode())

    async def write_many(self, payload: List[KV], ops: Optional[WriteOps]) -> int:
        if len(payload) == 0:
            return 0
        size_in_bytes = 0
        try:
            async with self._client.pipeline(transaction=True) as pipe:
                for kv in payload:
                    ttl = (
                        kv.expiration if kv.expiration else self._key_retention_seconds
                    )
                    string_val = ujson.dumps(kv.value)
                    this_size = len(string_val.encode())

                    if ops and ops.get("maxSize") and this_size > ops.get("maxSize"):  # type: ignore
                        warn(
                            "This value's size has exceeded the limit.",
                            size=this_size,
                            limit=ops.get("maxSize"),
                        )
                        raise QuotaError()

                    size_in_bytes += len(string_val.encode())

                    pipe.set(kv.key, string_val, ex=ttl)

                results = await pipe.execute()

            for result in results:
                if not result:
                    raise Exception(
                        "Attempt to write to redis resulted in a NON-OK response."
                    )

        except QuotaError:
            raise

        except redis.exceptions.TimeoutError as e:
            error("redis client timed out when trying to write", err=str(e))
            raise e

        except Exception as e:
            error("failed to write to redis", err=str(e))
            raise e

        log_fields = {"io": {"writes": size_in_bytes}}

        if ops and ops.get("baggage"):
            log_fields = {**log_fields, **ops.get("baggage")}  # type: ignore

        info("commited transaction", **log_fields)

        return size_in_bytes

    async def delete(self, keys: str) -> None:
        try:
            resp = await self._client.delete(keys)

            if not resp:
                raise Exception(
                    "Attempt to delete key in redis resulted in a NON-OK response"
                )
        except Exception as e:
            error(f"Failed to write to redis: {str(e)}")
            raise e

    async def close(self, reason: str = "") -> None:
        try:
            await self._client.close()
        except Exception as e:
            reason_string = f"Closing reason: {reason}, " if reason else ""
            error(f"{reason_string}Failed to close redis connection: {str(e)}")
